### PR TITLE
expression: implement vectorized evaluation for 'builtinCastRealAsStringSig'

### PR DESCRIPTION
**Pull request by tsthght**
### What problem does this PR solve? 
Implement vectorized evaluation for builtinCastRealAsStringSig, for #12106

### What is changed and how it works?
```
BenchmarkVectorizedBuiltinCastFunc/builtinCastRealAsStringSig-VecBuiltinFunc-8         	   10000	    226775 ns/op	   53040 B/op	    1660 allocs/op
BenchmarkVectorizedBuiltinCastFunc/builtinCastRealAsStringSig-NonVecBuiltinFunc-8      	    5000	    238973 ns/op	   53040 B/op	    1660 allocs/op
```
### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_cast_vec.go [hunk 1]
```diff
@@ -156,11 +156,48 @@ func (b *builtinCastTimeAsJSONSig) vecEvalJSON(input *chunk.Chunk, result *chunk
...
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalReal(b.ctx, input, buf); err != nil {
+		return err
+	}
+
+	bits := 64
+	if b.args[0].GetType().Tp == mysql.TypeFloat {
+		bits = 32
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec.go]: We also need the comment here.

File: expression/builtin_cast_vec.go [hunk 2]
```diff
@@ -156,11 +156,48 @@ func (b *builtinCastTimeAsJSONSig) vecEvalJSON(input *chunk.Chunk, result *chunk
...
+
+	f64s := buf.Float64s()
+	result.ReserveString(n)
+	sc := b.ctx.GetSessionVars().StmtCtx
+	for i, v := range f64s {
+		if buf.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		res, e := types.ProduceStrWithSpecifiedTp(strconv.FormatFloat(v, 'f', -1, bits), b.tp, sc, false)
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec.go]: s/ e/ err

File: expression/builtin_cast_vec.go [hunk 3]
```diff
@@ -156,11 +156,48 @@ func (b *builtinCastTimeAsJSONSig) vecEvalJSON(input *chunk.Chunk, result *chunk
...
+		if e != nil {
+			return e
+		}
+		str, b, e1 := padZeroForBinaryType(res, b.tp, b.ctx)
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec.go]: s/ str/ res
s/ b/ isNull
s/ e1/ err

We can define `var isNull bool` before line 181

File: expression/builtin_cast_vec_test.go [hunk 4]
```diff
@@ -37,6 +37,7 @@ var vecBuiltinCastCases = map[string][]vecExprBenchCase{
 		},
 		{retEvalType: types.ETString, childrenTypes: []types.EvalType{types.ETInt}},
 		{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETDatetime}},
+		{retEvalType: types.ETString, childrenTypes: []types.EvalType{types.ETReal}},
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec_test.go]: We may add a dataGenerator for this test case.

File: expression/builtin_cast_vec.go (hunk 1 above)
Comment (tsthght) [reply expression/builtin_cast_vec.go]: ok, i will add the comment here.

File: expression/builtin_cast_vec.go (hunk 2 above)
Comment (tsthght) [reply expression/builtin_cast_vec.go]: ok, i replace it

File: expression/builtin_cast_vec.go (hunk 3 above)
Comment (tsthght) [reply expression/builtin_cast_vec.go]: ok, i replace it

File: expression/builtin_cast_vec_test.go (hunk 4 above)
Comment (tsthght) [reply expression/builtin_cast_vec_test.go]: What's wrong with using the default dataGenerator?

Comment (XuHuaiyu) [reply expression/builtin_cast_vec_test.go]: Never mind, it's ok here

**Pull request merged**